activerecord/lib/active_record/database_configurations/database_config.rb

```diff
@@ -6,11 +6,51 @@ class DatabaseConfigurations
     # UrlConfig respectively. It will never return a DatabaseConfig object,
     # as this is the parent class for the types of database configuration objects.
     class DatabaseConfig # :nodoc:
+      include Mutex_m
+
       attr_reader :env_name, :spec_name
 
+      attr_accessor :schema_cache
+
       def initialize(env_name, spec_name)
+        super()
         @env_name = env_name
         @spec_name = spec_name
+        @pool = nil
+        @pool_pid = Process.pid
+      end
+
+      def disconnect!
+        discard_unowned_pool!
+
+        return unless @pool
+
+        synchronize do
+          return unless @pool
+
+          @pool.automatic_reconnect = false
+          @pool.disconnect!
+        end
+
+        nil
+      end
+
+      def connection_pool
+        discard_unowned_pool!
+
+        @pool || synchronize { @pool ||= ConnectionAdapters::ConnectionPool.new(self) }
+      end
+
+      def discard_unowned_pool!
+        return if @pool_pid == Process.pid
+
+        synchronize do
+          return if @pool_pid == Process.pid
```

Comment:
Already the `return if @pool_pid == Process.pid` has returned. why we need to check this again? 
Same as line #26 and #29.

Reply:
That's [double checked locking](https://en.wikipedia.org/wiki/Double-checked_locking).

While we're waiting to sync the mutex, another thread might have changed the state, so we check again to see if we still need to apply the change.

Reply:
```suggestion
     def discard_unowned_pool!
         synchronize do
             return if @pool_pid == Process.pid
             @pool&.discard!
             @pool = nil
             @pool_pid = Process.pid
           end
      end
```
Then might this alone be sufficent correct?

Reply:
Yes, but the whole point is to avoid synchronizing the mutex in most cases.